test/common/event/dispatcher_impl_test.cc

```diff
@@ -173,6 +173,49 @@ TEST_F(SchedulableCallbackImplTest, ScheduleChainingAndCancellation) {
   dispatcher_->run(Dispatcher::RunType::Block);
 }
 
+TEST_F(SchedulableCallbackImplTest, RescheduleNext) {
+  DispatcherImpl* dispatcher_impl = static_cast<DispatcherImpl*>(dispatcher_.get());
+  ReadyWatcher prepare_watcher;
+  evwatch_prepare_new(&dispatcher_impl->base(), onWatcherReady, &prepare_watcher);
+
+  ReadyWatcher watcher0;
+  createCallback([&]() {
+    watcher0.ready();
+    // Callback 1 was scheduled from the previous iteration, expect it to fire in the current
+    // iteration despite the attempt to reschedule.
+    callbacks_[1]->scheduleCallbackNextIteration();
+    // Callback 2 expected to execute next iteration because current because current called before
+    // next.
+    callbacks_[2]->scheduleCallbackCurrentIteration();
```

Comment:
I think you may need another pair of callbacks for completeness. I think this tests that the `scheduleCallbackCurrentIteration()` is working as expected by not rescheduling callback to the current iteration. However I do not think that it tests that the `scheduleCallbackNextIteration()` is working as expected, because if it is broken and does reschedule to the next iteration, the call order would remain the same.

Reply:
Callback 3 below may exercise the case you're referring to.

Notice the EXPECT_CALL(prepare_watcher, ready()); which is called at the beginning of each iteration and can be used to detect what runs in the first iteration vs second iteration of the loop.